core/trino-main/src/main/java/io/trino/operator/aggregation/partial/PartialAggregationController.java

```diff
@@ -57,15 +63,28 @@ public boolean isPartialAggregationDisabled()
         return partialAggregationDisabled;
     }
 
-    public synchronized void onFlush(long bytesProcessed, long rowsProcessed, long uniqueRowsProduced)
+    public synchronized void onFlush(long bytesProcessed, long rowsProcessed, OptionalLong uniqueRowsProduced)
     {
+        if (!partialAggregationDisabled && uniqueRowsProduced.isEmpty()) {
+            // when PA is re-enabled, ignore stats from disabled flushes
+            return;
+        }
+
         totalBytesProcessed += bytesProcessed;
         totalRowProcessed += rowsProcessed;
-        totalUniqueRowsProduced += uniqueRowsProduced;
+        uniqueRowsProduced.ifPresent(value -> totalUniqueRowsProduced += value);
 
         if (!partialAggregationDisabled && shouldDisablePartialAggregation()) {
             partialAggregationDisabled = true;
         }
+
+        if (partialAggregationDisabled
+                && totalBytesProcessed >= maxPartialMemory.toBytes() * ENABLE_AGGREGATION_BUFFER_SIZE_TO_INPUT_BYTES_FACTOR) {
```

Comment:
This is currently set to 1% of total data when PA is on. I think we can pay slight penalty in order to make the mechanism more resilient.

Reply:
if the cluster has 100 workers we will not be able to use RLE support because every 100th page will contain aggregated data. This will effectively disable RLE/dictionary support in `PagePartitioner` after partial aggregation.